doc/source/msg.rst

```diff
@@ -8,25 +8,108 @@ Message Protocol
 Overview
 --------
 
-The `bluesky.Msg` object is a `namedtuple` subclass which has the fields
+A *plan* is a sequence of atomic operations describing a data acquisition
+procedure. Each operation is represented by a ``bluesky.Msg`` ("message")
+object. A plan may be implemented as a simple list of messages:
+
+.. code-block:: python
+
+    from bluesky import Msg
+
+    # (Behold, the most boring data acquisition ever conducted!)
+    plan = [Msg('open_run'), Msg('close_run')]
+
+or as a generator the yields messages one at time:
+
+.. code-block:: python
+
+    def plan():
+        yield Msg('open_run')
+        yield Msg('close_run')
+
+The above examples are equivalent. For more sophisticated uses, the second one
+is more powerful, as it can incorporate loops, conditionals, adaptive logic ---
+generally any Python code.
+
+But, crucially, the plan code itself must not communicate with hardware.
+(You should never put ``epics.caput(...)`` in a plan!) Rather, each operation
+is represented by a ``Msg`` object that *describes* what should be done. This
+makes it safe to introspect the plan for error-checking, simulation, and
+visualization purposes --- without touching real hardware. For example, we
+could print each message in the plan like so:
+
+.. code-block:: python
+
+    plan = [Msg('open_run'), Msg('close_run')]
+
+    # a very, very simple 'plan simulator'
+    for msg in plan:
+        print(msg)
+
+A ``Msg`` has four members, accessible as attributes:
 
 - command
 - obj
 - args
 - kwargs
 
-``command`` must be one of a controlled list of commands, ``obj`` is the
-object to apply the command to and ``args`` and ``kwargs`` are arguments to
-the command.  Any ``args`` or ``kwargs`` not consumed by the run engine are
-passed through to the calls on the objects.
-
-The `RunEngine` has a registry which is used to dispatch the `Msg` objects
-based on the value of the `Msg.cmd`.  By default a basic set of commands are
-registered, but users can register their own functions to add custom commands.
+where ``command`` must be one of a controlled list of commands, ``obj`` is the
+object (i.e. Device) to apply the command to, if applicable, and ``args`` and
+``kwargs`` are arguments to the command.
+
+The execute the plan, the RunEngine consumes it, one message at a time.
+
+.. code-block:: python
+
+    def very_simple_run_engine(plan):
+        for msg in plan:
+            # Process the msg.
```

Comment:
should we include a link to https://nsls-ii.github.io/bluesky/run_engine.html

Reply:
👍

Reply:
@danielballan, I don't see it here. Did you push?

Reply:
done